### PR TITLE
Fix ipv6 address parsing with leading compressor

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
Fixes https://github.com/dotnet/corefx/issues/38634

There is a bug if address has got leading compressor.
Bug doesn't raise if we have trailing compressor because compressor index math  exists loop after one cycle.
Added a fix that skip expansion if from/to index calculation ends with same value, mean that zeros are already in the right place.

cc: @davidsh @wfurt @caesar-chen @rmkerr @karelz @scalablecory

I didn't found area owner for System.Net.Primitives https://github.com/dotnet/corefx/blob/d3942d4671919edb0cca6ddc1840190f524a809d/Documentation/project-docs/issue-guide.md#areas

### CODE REVIEW

**Comment by davidsh** [comment]
Thanks for the PR.  A couple of notes:

>cc: @davidsh @wfurt @caesar-chen @rmkerr @karelz @scalablecory

@caesar-chen and @rmkerr are no longer on the team.  Thus, they don't need to be added to PRs or issues.

Area owner for System.Net.Primitives is the same as all System.Net.*

What is the behavior of this issue on prior versions of .NET Core?  And IPv6 address parsing code is also present in System.Uri.  Can you see what it is does with IPv6 addresses similar to this? We should be making equivalent fixes if a fix is required.

**Comment by davidsh** [comment]
>I didn't found area owner for System.Net.Primitives https://github.com/dotnet/corefx/blob/d3942d4671919edb0cca6ddc1840190f524a809d/Documentation/project-docs/issue-guide.md#areas

I think your link to the area owners document is out-of-date. The current document in master is already updated.
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md

**Comment by MarcoRossignoli** [comment]
> I think your link to the area owners document is out-of-date. The current document in master is already updated.

Found the throuble I looking for area owner doc using github search and got an old version also if the link seem up-to-date mad in that doc there is an old link of owners.

image

**Comment by MarcoRossignoli** [comment]
Old versions
```cs
class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine(IPAddress.TryParse("::2:3:4:5:6:7:8", out var addr) ? addr.ToString() : "error");
        }
    }
```
```
D:\Users\marco_000\Downloads\TestIPV6\TestIPV6
λ dotnet run -f netcoreapp2.0
::

D:\Users\marco_000\Downloads\TestIPV6\TestIPV6
λ dotnet run -f netcoreapp2.1
::

D:\Users\marco_000\Downloads\TestIPV6\TestIPV6
λ dotnet run -f netcoreapp2.2
::

D:\Users\marco_000\Downloads\TestIPV6\TestIPV6
λ dotnet run -f netcoreapp3.0
::

D:\Users\marco_000\Downloads\TestIPV6\TestIPV6
λ
```

**Comment by MarcoRossignoli** [comment]
> And IPv6 address parsing code is also present in System.Uri. Can you see what it is does with IPv6 addresses similar to this? We should be making equivalent fixes if a fix is required.

@davidsh I cannot open funcional tests with VS 2019 preview(cc: @ViktorHofer)

image

Btw `System.Private.Uri` uses shared common ipv6 parser updated by this PR.
https://github.com/dotnet/corefx/blob/a7e0c8d9e14a4f814a7e08ddfd0316353c37dd95/src/System.Private.Uri/src/System.Private.Uri.csproj#L25

**Comment by davidsh** [comment]
>@davidsh I cannot open funcional tests with VS 2019 preview(cc: @ViktorHofer)

You can also use command line to build and run tests, i.e. "msbuild /t:rebuildandtest" in the src/System.Private.Uri/tests/FunctionalTests.

>Btw System.Private.Uri uses shared common ipv6 parser updated by this PR.

Not all code is in Common.  Here is an example of code that isn't:
https://github.com/dotnet/corefx/blob/master/src/System.Private.Uri/src/System/IPv6AddressHelper.cs

You should verify functionality by adding some similar tests to System.Private.Uri here:
https://github.com/dotnet/corefx/blob/master/src/System.Private.Uri/tests/FunctionalTests/UriIpHostTest.cs

See recent PR #37734 which fixed other IPv6 address parsing problems across both IPAddress and System.Uri classes.

**Comment by MarcoRossignoli** [comment]
> You can also use command line to build and run tests, i.e. "msbuild /t:rebuildandtest" in the src/System.Private.Uri/tests/FunctionalTests directory.

Yep tests run I'll add some tests asap thank's!

**Comment by stephentoub** [comment]
>  cannot open funcional tests with VS 2019 preview

@ViktorHofer, I'm hitting this same issue, with System.Net.NameResolution.  Any ideas?

File: src/Common/src/System/Net/IPv6AddressHelper.Common.cs [hunk 1]
```diff
@@ -411,6 +411,13 @@ internal static void Parse(ReadOnlySpan<char> address, Span<ushort> numbers, int
                 int toIndex = NumberOfLabels - 1;
                 int fromIndex = index - 1;
 
+                // if fromIndex and toIndex are the same 0 bits are already in the correct place
```
Comment (danmosemsft) [inline src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: does this mean `if fromIndex and toIndex are the same 0, then bits are already in the correct place` or something else? could be clearer.

Comment (MarcoRossignoli) [reply src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: The meaning is that if from and to are same "0 bits"(compressed part) is already in correct place and doesn't need to moved up in the array position(called expanded in comment above) so maybe I need to place a comma before "0 bits".
Algorithm parse 16bits group in a sequence way and bookmark when a compressor(::) found.
If we have a compressor in the address means that at the end of parsed numbers we have the "0 compressor value" and we need to move in correct place up to arary index.
If calculation of fromIndex and ToIndex is the same mean that "0 value" is already in the correct position and we can return.

Comment (MarcoRossignoli) [reply src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: I've updated the comment, let me know if it's more clear.

**Comment by MarcoRossignoli** [comment]
@davidsh PTAL 
Added tests for `System.Private.Uri` thank's for pointing me right test file.
I followed the code and it jumps inside common address helper here https://github.com/dotnet/corefx/blob/master/src/System.Private.Uri/src/System/IPv6AddressHelper.cs#L20

File: src/Common/src/System/Net/IPv6AddressHelper.Common.cs [hunk 2]
```diff
@@ -411,6 +411,13 @@ internal static void Parse(ReadOnlySpan<char> address, Span<ushort> numbers, int
                 int toIndex = NumberOfLabels - 1;
                 int fromIndex = index - 1;
 
+                // if fromIndex and toIndex are the same mean that "zero bits" are already in the correct place
+                // it happens for leading and trailing compression
+                if (fromIndex == toIndex)
+                {
+                    return;
+                }
```
Comment (stephentoub) [inline src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: Nit: could you instead make this:
```C#
if (fromIndex != toIndex)
{
    for (int i = ...)
    {
        ...
    }
}
```
The `return` here is a bit hidden.

File: src/Common/src/System/Net/IPv6AddressHelper.Common.cs [hunk 3]
```diff
@@ -411,6 +411,13 @@ internal static void Parse(ReadOnlySpan<char> address, Span<ushort> numbers, int
                 int toIndex = NumberOfLabels - 1;
                 int fromIndex = index - 1;
 
+                // if fromIndex and toIndex are the same mean that "zero bits" are already in the correct place
```
Comment (stephentoub) [inline src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: Nit:
"if fromIndex and toIndex are the same mean that" =>
"If fromIndex and toIndex are the same, it means that"

File: src/Common/src/System/Net/IPv6AddressHelper.Common.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: I did it in first version 😄 ok.

File: src/Common/src/System/Net/IPv6AddressHelper.Common.cs (hunk 3 above)
Comment (MarcoRossignoli) [reply src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: ok thank's

File: src/Common/src/System/Net/IPv6AddressHelper.Common.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: done

File: src/Common/src/System/Net/IPv6AddressHelper.Common.cs (hunk 3 above)
Comment (MarcoRossignoli) [reply src/Common/src/System/Net/IPv6AddressHelper.Common.cs]: done

**Comment by davidsh** [comment]
I'd like to pause on this PR for now.

I've done some research on whether or not IPv6 addresses such as "::2:3:4:5:6:7" are actually valid IPv6 addresses.

According to WinRT Windows.Networking.Hostname API parsing, it is not valid.  The following code throws an exception "The parameter is incorrect".

I think the premise of the issue raised by #38634 is not correct.  I will follow-up after I've discussed with the IETF experts for the IPv6 RFC.

**Comment by wfurt** [comment]
::1 is certainly valid. Did we not process it correctly before?

**Comment by davidsh** [comment]
>::1 is certainly valid. Did we not process it correctly before?

Yes, "::1" is valid.  This PR is specifically about the case where there are 7 distinct groups of numbers and one compressor which is a leading compressor, i.e. 

"::2:3:4:5:6:7:8".  The debate is about whether "::" can represent at least a single "0" (or multiple)  *OR* whether it can only represents two groups of zeroes (or more than that).  I'm waiting to hear back from IETF rep.

**Comment by wfurt** [comment]
It would be good to get the feedback. I did quick test with inet_ntop() and inet_pton() on Linux and they process "::2:3:4:5:6:7:8" as valid string expanded to "0:2:3:4:5:6:7:8"

**Comment by davidsh** [comment]
>It would be good to get the feedback. I did quick test with inet_ntop() and inet_pton() on Linux and they process "::2:3:4:5:6:7:8" as valid string expanded to "0:2:3:4:5:6:7:8"

@wfurt Thanks for the additional validations with the Linux native APIs!

After discussions with the IETF rep and the Windows team, the RFC is correct.  So, the fixes for IPAddress class are correct. It turns out there is a separate bug for the WinRT Windows.Networking.HostName class that will be filed separately.

**Comment by MarcoRossignoli** [comment]
Unrelated fails

**Pull request merged**